### ISSUE TITLE
v0.0.18 Most entities do not update/change.

### ISSUE THREAD

**Issue by lackr007** [mikey0000/Mammotion-HA#63]
### The problem

When the device is initially added the entities sensors values are correct, however they never update after that. Mowing Speed and Blade Height showing unavailable. Area, Activity mode, Elapsed time, L1 Satellites, L2 Satellites, Progress, Time left, and Total time not updating.

### What Mammotion mower are you using?

Luba 2

### What version of the Mammotion intergration are you using?

v0.0.18

### What version of Home Assistant Core are you running?

core- 2024.8.1

**Comment by lackr007** [comment]
Here are some the logs that pertain to mammotion. FYI, In my setup I have the bluetooth proxy ESP32 mounted inside my LUBA 2 (powered by LUBA), I have 3 outside wireless APs that keep it connected to wifi. I am running HA automations that open/close garage doors so it can travel to/from my back yard and front yard, my problem is my automations references some of the entities from this mammotion integration. Ever since version 0.0.5 my automations haven't worked correctly, but version 0.0.5 the entities worked flawlessly on my setup. Thanks for all the hard work on this integration, truly appreciated!


This error originated from a custom integration.

Logger: homeassistant
Source: custom_components/mammotion/sensor.py:217
integration: Mammotion
First occurred: 9:24:37 AM (229 occurrences)
Last logged: 2:12:38 PM

Error doing job: Task exception was never retrieved (None)
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 258, in _handle_refresh_interval
    await self._async_refresh(log_failures=True, scheduled=True)
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 453, in _async_refresh
    self.async_update_listeners()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 168, in async_update_listeners
    update_callback()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 533, in _handle_coordinator_update
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1005, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1130, in _async_write_ha_state
    self.__async_calculate_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1067, in __async_calculate_state
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1011, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 542, in state
    value = self.native_value
            ^^^^^^^^^^^^^^^^^
  File "/config/custom_components/mammotion/sensor.py", line 217, in native_value
    return DistanceConverter.convert(current_value, LENGTH_UNITS[unit], UnitOfLength.FEET)
                                                    ~~~~~~~~~~~~^^^^^^
TypeError: 'set' object is not subscriptable






Logger: homeassistant.components.sensor
Source: helpers/entity_platform.py:598
integration: Sensor (documentation, issues)
First occurred: 9:23:36 AM (1 occurrences)
Last logged: 9:23:36 AM

Error adding entity sensor.luba_vscvr9dm_blade_height for domain sensor with platform mammotion
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 598, in _async_add_entities
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 912, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1366, in add_to_platform_finish
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1005, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1130, in _async_write_ha_state
    self.__async_calculate_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1067, in __async_calculate_state
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1011, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 542, in state
    value = self.native_value
            ^^^^^^^^^^^^^^^^^
  File "/config/custom_components/mammotion/sensor.py", line 217, in native_value
    return DistanceConverter.convert(current_value, LENGTH_UNITS[unit], UnitOfLength.FEET)
                                                    ~~~~~~~~~~~~^^^^^^
TypeError: 'set' object is not subscriptable





Logger: homeassistant.components.sensor
Source: helpers/entity_platform.py:598
integration: Sensor (documentation, issues)
First occurred: 9:23:36 AM (1 occurrences)
Last logged: 9:23:36 AM

Error adding entity sensor.luba_vscvr9dm_mowing_speed for domain sensor with platform mammotion
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 598, in _async_add_entities
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 912, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1366, in add_to_platform_finish
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1005, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1130, in _async_write_ha_state
    self.__async_calculate_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1067, in __async_calculate_state
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1011, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 542, in state
    value = self.native_value
            ^^^^^^^^^^^^^^^^^
  File "/config/custom_components/mammotion/sensor.py", line 219, in native_value
    return SpeedConverter.convert(current_value, SPEED_UNITS[unit], UnitOfSpeed.FEET_PER_SECOND)
                                                 ~~~~~~~~~~~^^^^^^
TypeError: 'set' object is not subscriptable






Logger: homeassistant
Source: /usr/src/homeassistant/homeassistant/runner.py:148
First occurred: 9:29:14 AM (52 occurrences)
Last logged: 10:30:54 AM

Error doing job: Task exception was never retrieved (None)
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/pymammotion/mammotion/devices/mammotion.py", line 541, in _notification_handler
    self._update_raw_data(data)
  File "/usr/local/lib/python3.12/site-packages/pymammotion/mammotion/devices/mammotion.py", line 197, in _update_raw_data
    tmp_msg = LubaMsg().parse(data)
              ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 759, in parse
    value = self._postprocess_single(
            ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 701, in _postprocess_single
    fmt = _pack_fmt(meta.proto_type)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 281, in _pack_fmt
    return {
           ^
KeyError: 'enum'






Logger: homeassistant
Source: /usr/src/homeassistant/homeassistant/runner.py:148
First occurred: 9:31:45 AM (131 occurrences)
Last logged: 10:35:08 AM

Error doing job: Task exception was never retrieved (None)
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/pymammotion/mammotion/devices/mammotion.py", line 541, in _notification_handler
    self._update_raw_data(data)
  File "/usr/local/lib/python3.12/site-packages/pymammotion/mammotion/devices/mammotion.py", line 197, in _update_raw_data
    tmp_msg = LubaMsg().parse(data)
              ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 759, in parse
    value = self._postprocess_single(
            ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 718, in _postprocess_single
    value = cls().parse(value)
            ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 759, in parse
    value = self._postprocess_single(
            ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 718, in _postprocess_single
    value = cls().parse(value)
            ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 759, in parse
    value = self._postprocess_single(
            ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 718, in _postprocess_single
    value = cls().parse(value)
            ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 752, in parse
    decoded, pos = decode_varint(parsed.value, pos)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 390, in decode_varint
    b = buffer[pos]
        ~~~~~~^^^^^
IndexError: index out of range







Logger: homeassistant
Source: /usr/src/homeassistant/homeassistant/runner.py:148
First occurred: 10:39:14 AM (59 occurrences)
Last logged: 10:42:08 AM

Error doing job: Task exception was never retrieved (None)
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/pymammotion/mammotion/devices/mammotion.py", line 541, in _notification_handler
    self._update_raw_data(data)
  File "/usr/local/lib/python3.12/site-packages/pymammotion/mammotion/devices/mammotion.py", line 197, in _update_raw_data
    tmp_msg = LubaMsg().parse(data)
              ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 759, in parse
    value = self._postprocess_single(
            ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 718, in _postprocess_single
    value = cls().parse(value)
            ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 759, in parse
    value = self._postprocess_single(
            ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 718, in _postprocess_single
    value = cls().parse(value)
            ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 754, in parse
    decoded = self._postprocess_single(
              ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 702, in _postprocess_single
    value = struct.unpack(fmt, value)[0]
            ^^^^^^^^^^^^^^^^^^^^^^^^^
struct.error: unpack requires a buffer of 8 bytes

**Comment by mikey0000** [comment]
Oh your American! I can see why, give me a tic and I'll fix that for you!

**Comment by mikey0000** [comment]
ok try with 0.0.19

**Comment by lackr007** [comment]
Haha... yes, I’m American. I had a hunch it might be a unit conversion issue! Our units of measure always remind me of one of my favorite SNL skits...  Thanks again! 😄 https://www.youtube.com/watch?v=JYqfVE-fykk

**Comment by mikey0000** [comment]
That's hilarious! Hey I'm going to fix up the entities for us measurements so 0.0.20 coming shortly.

**Comment by mikey0000** [comment]
Hey fixed it all properly now, re-download 0.0.19

**Comment by mikey0000** [comment]
you will need to go into the entity and set the unit you want for blade height and speed.
